compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -372,23 +404,63 @@ public void visit(BLangSimpleVariableDef varDefNode) {
      * @return assignment statement created
      */
     private BLangAssignment createAssignment(BLangSimpleVariableDef varDefNode) {
-        // Create a map for the function or block node.
-        BLangBlockStmt blockStmt = (BLangBlockStmt) env.node;
-        if (blockStmt.mapSymbol == null) {
-            // create a map symbol
-            blockStmt.mapSymbol = createMapSymbol("$map$block$" + blockClosureMapCount, env);
-        }
+        BVarSymbol mapSymbol = createMapSymbolIfAbsent(env.node, blockClosureMapCount);
 
         // Add the variable to the created map.
-        BLangIndexBasedAccess accessExpr = ASTBuilderUtil.createIndexBasesAccessExpr(varDefNode.pos, varDefNode.type,
-                blockStmt.mapSymbol, ASTBuilderUtil.createLiteral(varDefNode.pos, symTable.stringType,
-                        varDefNode.var.name.value));
-        accessExpr.type = ((BMapType) blockStmt.mapSymbol.type).constraint;
+        BLangIndexBasedAccess accessExpr =
+                ASTBuilderUtil.createIndexBasesAccessExpr(varDefNode.pos, varDefNode.type, mapSymbol,
+                                                          ASTBuilderUtil
+                                                                  .createLiteral(varDefNode.pos, symTable.stringType,
+                                                                                 varDefNode.var.name.value));
+        accessExpr.type = ((BMapType) mapSymbol.type).constraint;
         accessExpr.lhsVar = true;
         // Written to: 'map["x"] = 8'.
         return ASTBuilderUtil.createAssignmentStmt(varDefNode.pos, accessExpr, varDefNode.var.expr);
     }
 
+    private BVarSymbol createMapSymbolIfAbsent(BLangNode node, int closureMapCount) {
+        if (node.getKind() == NodeKind.BLOCK_FUNCTION_BODY) {
+            return createMapSymbolIfAbsent((BLangBlockFunctionBody) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.BLOCK) {
+            return createMapSymbolIfAbsent((BLangBlockStmt) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.FUNCTION) {
+            return createMapSymbolIfAbsent((BLangFunction) node, closureMapCount);
+        }
+        return null;
+    }
+
+    private BVarSymbol createMapSymbolIfAbsent(BLangBlockFunctionBody body, int closureMapCount) {
```

Comment:
Can ```getOrCreateMapSymbol``` be a better name.

Reply:
Sort of gives the idea that the method has multiple responsibilities doesn't it?